### PR TITLE
Caching OpenSearch builds

### PR DESCRIPTION
**Pull request by nhtruong**
### Description
Reduce workflows runtime
### Issues Resolved
closes https://github.com/opensearch-project/opensearch-ruby/issues/144

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: .github/workflows/test-unreleased.yml [hunk 1]
```diff
@@ -14,34 +14,50 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        entry:
-          - { opensearch_ref: '1.x' }
-          - { opensearch_ref: '2.x' }
-          - { opensearch_ref: '2.0' }
-          - { opensearch_ref: 'main' }
+        opensearch_ref: [ '1.x', '2.x', '2.0', 'main' ]
+        ruby_version: 3.1
```
Comment (harshavamsi) [inline .github/workflows/test-unreleased.yml:18]: Do we want to make this `3.2` or add `3.2` to the matrix now that it is released?

File: .github/workflows/test-unreleased.yml [hunk 2]
```diff
@@ -64,3 +80,10 @@ jobs:
         run: cd opensearch-dsl && bundle exec rake test:all
       - name: opensearch-aws-sigv4
         run: cd opensearch-aws-sigv4 && bundle exec rake test:all
+
+      - name: Save server logs
+        if: failure()
+        uses: actions/upload-artifact@v3
+        with:
+          name: opensearch-logs-${{ matrix.opensearch_ref }}-python-${{ matrix.ruby_version }}
```
Comment (harshavamsi) [inline .github/workflows/test-unreleased.yml:88]:
```suggestion
          name: opensearch-logs-${{ matrix.opensearch_ref }}-ruby-${{ matrix.ruby_version }}
```

**Review by harshavamsi — CHANGES_REQUESTED** [review]

File: .github/workflows/test-unreleased.yml (hunk 1 above)
Comment (nhtruong) [reply .github/workflows/test-unreleased.yml:18]: I don't think it matters. The goal of this workflow is to test the unreleased OS. We have other workflows that test Ruby 3.2

**Review by harshavamsi — APPROVED** [review]

**Review by VachaShah — APPROVED** [review]

**Comment by VachaShah** [comment]
@nhtruong The integration unreleased workflow is failing after this change https://github.com/opensearch-project/opensearch-ruby/actions/runs/4247268700. Can you please take a look? Thank you!

**Pull request merged**